public/app.js

```diff
@@ -0,0 +1,84 @@
+// GLOBAL APP CONTROLLER
+class Controller {
+    static setupEventListeners() {
+        let DOMstrings = UIController.getDOMstrings();
+        document
+            .querySelector(DOMstrings.logoutBtn)
+            .addEventListener("click", AuthController.logout);
+        document
+            .querySelector(DOMstrings.codingProgressLinkSelector)
+            .addEventListener("click", Controller.navigateToCodingProgress);
+        document
+            .querySelector(DOMstrings.projectMenu)
+            .addEventListener("click", Controller.navigateToSelectedProject);
+    }
+
+    static resetUI() {
+        let DOMstrings = UIController.getDOMstrings();
+        document.querySelector(DOMstrings.codingProgressContainer).innerHTML = "";
+        document.querySelector(DOMstrings.graphContainer).innerHTML = "";
+    }
+
+    static detachSnapshotListener() {
+        let DOMstrings = UIController.getDOMstrings(),
+            hasListener = false,
+            nodeList = document.querySelectorAll(DOMstrings.dropdownItem);
+        if (nodeList) {
+            // Makes sure all the snapshot listeners of the projects' traffic data are detached
+            Array.prototype.forEach.call(nodeList, obj => {
+                let project = obj.innerText;
```

Comment:
This reads the project names out of the DOM, which creates a backward dependency that we were trying to avoid by restructuring the project like this. I think it would be better to have the watch functions in DataController return the unsubscribe function, save them to a variable in this class, then execute those functions and remove them from the variable when detaching.

Reply:
resolved